Code for QUIC support

helkoulak opened the pull request.

helkoulak committed:
The PR contribute the following code:

- Cryptography code to support QUIC for header and packet protection.
- Adds function extract_key_pair() to extract ED25519 private and if available public key values from a PKCS#8 DER formatted key.
- Few bug fixes for the rustls provider.
- The test suite to test WolfCrypt provider against rustls version 0.23.35.
- Adds related workflows for building and testing.

julek-wolfssl commented on rustls-wolfcrypt-provider/src/hkdf.rs:
```diff
@@ -147,6 +147,8 @@ impl tls13::HkdfExpander for WolfHkdfExpander {
 }
 
+
+
```
Remove unnecessary file mod.

julek-wolfssl commented on rustls-wolfcrypt-provider/src/types/mod.rs:
```diff
@@ -123,6 +123,8 @@ macro_rules! define_foreign_type_with_copy {
     };
 }
 
+
+
```
Remove unnecessary file mod.

julek-wolfssl commented on rustls-wolfcrypt-provider/src/error.rs:
```diff
@@ -110,6 +110,8 @@ pub fn check_if_greater_than_zero(ret: i32) -> WCResult {
 }
 
+
+
```
Remove unnecessary file mod.

julek-wolfssl commented on rustls-wolfcrypt-provider/Cargo.toml:
```diff
@@ -33,6 +33,7 @@ lazy_static = "1.5.0"
 hex-literal = "0.4.1"
 
 
+
```
Remove unnecessary file mod.

julek-wolfssl reviewed (changes requested):
Good stuff. Does quic get tested in the current CI workflows?

gasbytes reviewed (changes requested):
Nice work.
There seems to be some formatting issues that the CI detected, please fix those.
Also address the github copilot reports (mainly some unwraps that need to be refactored to handle errors).
Run clippy locally before committing too, the CI should detect some lints with the newest version.
Thank you.

helkoulak commented:
> Good stuff. Does quic get tested in the current CI workflows?

Thank you. maybe we need to add --features quic in order for quic code to be compiled. 

> Nice work. There seems to be some formatting issues that the CI detected, please fix those. Also address the github copilot reports (mainly some unwraps that need to be refactored to handle errors). Run clippy locally before committing too, the CI should detect some lints with the newest version. Thank you.

Thank you. I have corrected the mentioned issues. Hopefully CI tests will runs without errors

gasbytes commented:
Hello @helkoulak, there are still some clippy reports in the CI/CD apparently, it would be great if you could try and fix them.

> Thank you. maybe we need to add --features quic in order for quic code to be compiled.

Yes, please feel free to update the current workflows to test that too.

helkoulak commented:
> Hello @helkoulak, there are still some clippy reports in the CI/CD apparently, it would be great if you could try and fix them.
>  No problem, I will fix them ASAP.
> > Thank you. maybe we need to add --features quic in order for quic code to be compiled.
> 
> Yes, please feel free to update the current workflows to test that too.
Thank you, will do that ASAP

julek-wolfssl commented on .github/workflows/ubuntu-build.yml:
```diff
@@ -57,6 +57,34 @@ jobs:
+          rm -rf rustls_v0.23.35_test_files
+          cargo test -p tests --test all_suites --all-features
+          cd ..
+          rm -rf rustlsv0.23.35-test-workspace
```
Let's split this up into steps to make it easier to understand what failed for future devs.

helkoulak replied:
Done

julek-wolfssl commented on .github/workflows/ubuntu-build.yml:
```diff
@@ -57,6 +57,34 @@ jobs:
+          cd ..
+          git clone https://github.com/helkoulak/rustls-wolfcrypt-provider.git
+          cd rustls-wolfcrypt-provider/
+          git checkout quic-support
```
Use https://github.com/marketplace/actions/checkout to fetch repos.

helkoulak replied:
Done

julek-wolfssl commented on .github/workflows/ubuntu-build.yml:
```diff
@@ -57,6 +57,34 @@ jobs:
+          cd ../rustls-wolfcrypt-provider/
+          cargo build --all-features --release
+          cd ../..
+          git clone https://github.com/helkoulak/rustls_v0.23.35_test_files.git
```
Place these under `.github/workflows/rustls` or a similar path. I don't see a need to have this in a dedicated repo.

julek-wolfssl replied:
If I understand correctly, you are using these files to make modifications to `rustls/rustls`? If yes, then I would prefer to use a patch file for this. The patch file should be applied with `patch -p1 < <path/to/patch/file>` in the `rustls/rustls` root dir.

helkoulak replied:
The idea here is to build a workspace that contains three packages, namely rustls, wolfcrypt-provider, and tests package. It is done this way to break the cyclic dependency between rustls and wolfcrypt-provider. The package tests is created from a modified copy of rustls/tests plus other files like build.rs and own Cargo.toml file.

Using a patch file will make in my opinion the script more difficult to understand. In addition to that, in each rustls version new tests are added and some are removed, so this makes the patch file valid for only a specific version of rustls.

gasbytes replied:
Pushed a commit (ca65a95) that addresses this.

julek-wolfssl commented on .github/workflows/ubuntu-build.yml:
The same comments apply to `macos-build.yml`.

julek-wolfssl replied:
I would consider moving the `rustls` tests into a separate workflow.

helkoulak replied:
Done

julek-wolfssl commented on rustls-wolfcrypt-provider/src/lib.rs:
```diff
@@ -50,7 +53,7 @@ type SigningAlgorithms = Vec<Box<SigningKeyFn>>;
 /*
  * Crypto provider struct that we populate with our own crypto backend (wolfcrypt).
  * */
-pub fn provider() -> CryptoProvider {
+pub fn default_provider() -> CryptoProvider {
```
Why is this necessary? Is this overloading something in rustls?

helkoulak replied:
Both providers, namely ring and aws-lc-rs, have it as default_provider(). So to make wolfcrypt more compatible with them and to do less manual work in the future testing rustls api, I have changed it this why.

julek-wolfssl commented on rustls-wolfcrypt-provider/src/sign/eddsa.rs:
```diff
@@ -20,65 +20,139 @@ pub struct Ed25519PrivateKey {
+        }
+
+        // Skip version (3 bytes), algorithm ID sequence (0x30 + length encoding bytes + 5 ID bytes),
+        skip_bytes += 3 + 7;
+        key_sub_slice = input_key.get(skip_bytes..).unwrap();
+
+        // Check if next bytes are 0x04, 0x22, 0x04, 0x20
+        if !matches!(
+            key_sub_slice,
+            [TAG_OCTET_SEQUENCE, 0x22, TAG_OCTET_SEQUENCE, 0x20, ..]
+        ) {
+            return Err(rustls::Error::General(
+                "Faulty PKCS#8 ED25519 private key structure".into(),
+            ));
+        }
+
+        // Copy private key value
+        skip_bytes += 4;
+        key_sub_slice = input_key.get(skip_bytes..).unwrap();
+        private_key_raw.copy_from_slice(&key_sub_slice[..ED25519_KEY_SIZE as usize]);
+        skip_bytes += ED25519_KEY_SIZE as usize;
+        key_sub_slice = input_key.get(skip_bytes..).unwrap();
+
+        // Check if optional SET OF attributes exists and skip
+        if key_sub_slice.first() == Some(&TAG_OPTIONAL_SET_OF_ATTRIBUTES) {
+            skip_bytes += (2 + (key_sub_slice[1])) as usize;
+            key_sub_slice = input_key.get(skip_bytes..).unwrap();
+        }
+
+        // Check if optional public key value exists. If exists, skip tag, length encoding byte,
+        // and bits-used byte
+        if key_sub_slice.first() == Some(&TAG_OPTIONAL_PUBLIC_KEY_BIT_STRING) {
+            public_key_raw.copy_from_slice(&key_sub_slice[3..(3 + ED25519_KEY_SIZE as usize)]);
+            Ok((private_key_raw, Some(public_key_raw)))
+        } else {
+            Ok((private_key_raw, None))
+        }
+    }
+}
+
```
I would prefer to use the ASN parsing in wolfcrypt. You should use `wc_Ed25519PrivateKeyDecode` which will pick up both priv and pub keys. Then `wc_ed25519_export_key` can be used to export the raw values. Only downside I see is that `privKeySet` and `pubKeySet` are not checked in `wc_ed25519_export_key`. Feel free to apply the following patch to wolfSSL. I'll submit a PR to wolfssl in a second.

```
diff --git a/wolfcrypt/src/ed25519.c b/wolfcrypt/src/ed25519.c
index a03efb5603..ff59e07310 100644
--- a/wolfcrypt/src/ed25519.c
+++ b/wolfcrypt/src/ed25519.c
@@ -1127,6 +1127,9 @@ int wc_ed25519_export_public(const ed25519_key* key, byte* out, word32* outLen)
         return BUFFER_E;
     }
 
+    if (!key->pubKeySet)
+        return PUBLIC_KEY_E;
+
     *outLen = ED25519_PUB_KEY_SIZE;
     XMEMCPY(out, key->p, ED25519_PUB_KEY_SIZE);
 
@@ -1368,7 +1371,7 @@ int wc_ed25519_export_private_only(const ed25519_key* key, byte* out, word32* ou
 int wc_ed25519_export_private(const ed25519_key* key, byte* out, word32* outLen)
 {
     /* sanity checks on arguments */
-    if (key == NULL || out == NULL || outLen == NULL)
+    if (key == NULL || !key->privKeySet || out == NULL || outLen == NULL)
         return BAD_FUNC_ARG;
 
     if (*outLen < ED25519_PRV_KEY_SIZE) {
@@ -1398,6 +1401,8 @@ int wc_ed25519_export_key(const ed25519_key* key,
 
     /* export public part */
     ret = wc_ed25519_export_public(key, pub, pubSz);
+    if (ret == PUBLIC_KEY_E)
+        ret = 0; /* ignore no public key */
 
     return ret;
 }
```

helkoulak replied:
I have applied the patch file provided but unfortunately with no luck. As I have explained earlier in our email conversations, the problem lies in the function wc_Ed25519PrivateKeyDecode. It is unable to parse/recognize the public key part found in a PKCS8 DER formatted key pair. Luckily I was able to debug and record where the error exactly happens. I will send it to you per Email.

helkoulak replied:
And the function extract_key_pair was the workaround

gasbytes replied:
This is going to be addressed by me in a following PR.
Now that this patch was introduced at commit dd35f10b579e1f389bf76fea4d51171dc8fb1b7b (https://github.com/wolfssl/wolfssl/pull/9553).

The following PR will contain these changes (using wc_Ed25519PrivateKeyDecode instead) + a release that pulls 5.9.1 which contains those changes that @julek-wolfssl worked on.

julek-wolfssl reviewed (changes requested):
Great work. Still need to dive into `src/aead`.

julek-wolfssl commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
```diff
@@ -0,0 +1,1713 @@
+impl Eq for HPAlgorithm {}
+
+/// AES-128.
+pub static AES_128: HPAlgorithm = HPAlgorithm {
+    key_len: AES_128_KEY_LEN,
+    hp_mask: generate_mask_aes,
+    id: HPAlgorithmID::Aes128,
+    init: init_hp_aes_cipher,
+};
+
+/// AES-256.
+pub static AES_256: HPAlgorithm = HPAlgorithm {
+    key_len: AES_256_KEY_LEN,
+    hp_mask: generate_mask_aes,
+    id: HPAlgorithmID::Aes256,
+    init: init_hp_aes_cipher,
+};
```
Should these be `const`?

helkoulak replied:
I think its correct as it is. Because here we look for a program life time validity rather than fixed values

julek-wolfssl commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
```diff
@@ -0,0 +1,1713 @@
+
+        // The `unwrap()` will not panic because `new_mask` returns a
+        // non-empty result.
+        let (first_mask, pn_mask) = mask.split_first().unwrap();
```
Let's avoid `unwrap`. I see rust has some ways to use `unwrap` without panicking. libs shouldn't take down an app. Maybe `unwrap_or_else` is appropriate?

julek-wolfssl commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
```diff
@@ -0,0 +1,1713 @@
+
+impl Eq for AeadAlgorithm {}
+
+/// AES-128 in GCM mode with 128-bit tags and 96 bit nonces.
+pub static AES_128_GCM: AeadAlgorithm = AeadAlgorithm {
+    init: init_aes_gcm_cipher,
+    encrypt: encrypt_aes_gcm,
+    decrypt: decrypt_aes_gcm,
+    key_len: AES_128_KEY_LEN,
+    id: PacketKeyAlgorithmID::Aes128Gcm,
+};
+
+/// AES-256 in GCM mode with 128-bit tags and 96 bit nonces.
+pub static AES_256_GCM: AeadAlgorithm = AeadAlgorithm {
+    init: init_aes_gcm_cipher,
+    encrypt: encrypt_aes_gcm,
+    decrypt: decrypt_aes_gcm,
+    key_len: AES_256_KEY_LEN,
+    id: PacketKeyAlgorithmID::Aes256Gcm,
+};
```
`const`?

helkoulak replied:
I think it is correct this way

julek-wolfssl commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
Let's avoid `unwrap`. I see rust has some ways to use `unwrap` without panicking. libs shouldn't take down an app. None of the places where `unwrap` is used are places where failure is catastrophic. Instead an error should be returned and user should be allowed to handle it that way . Maybe `unwrap_or_else` is appropriate?

julek-wolfssl replied:
Fine for tests though.

helkoulak replied:
I have replaced unwrap() completely with functions like ? and map_err to propagate errors instead of panicking.

gasbytes commented on .github/workflows/ubuntu-build.yml:
One thing that I noticed is that you compile and run the tests with `--all-features`. I understand that with `wolfcrypt-provider` feature only the wolfcrypt provider is going to be used against the rustls testsuite, which is great, but with `--all-features` `awc-lc-rs` and ring get compiled too even if they are not used.
Is it possible to remove them? Or are they hardcoded dependencies used in some way in the testsuite?

I think the command to run the testsuite with only the wolfcrypt provider (which is already compiled in) is:
`cargo test -p tests --test all_suites --features wolfcrypt-provider,tls12 --no-default-features`
or something similar.

helkoulak replied:
You are right. No need to use --all-features option. The tests run successfully using your suggested command with a few features added and it should be like this
cargo test -p tests --test all_suites --features wolfcrypt-provider,tls12,fips,zlib,prefer-post-quantum,logging --no-default-features

gasbytes commented on .github/workflows/ubuntu-build.yml:
Me and Juliusz were also thinking that it might be worth to add a feature to print the current provider being used via cargo, since you added the configuration option `wolfcrypt-provider`.
And add that step before running the testsuite, by grepping the output from stdout and confirming that we are running the full testsuite against the `wolfcrypt-provider` only.
That would be great.

helkoulak replied:
If I understand correctly, this is already done. The command that runs the tests targets only the runner file all_test_suites.rs. And in this runner file you have macros that are annotated with #[cfg(feature = "wolfcrypt-provider")] and print the sentence tests_with_wolfcrypt_. So as per my understanding, there is no way the tests will run against other providers than wolfcrypt-provider. Or did I miss something here?

gasbytes commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
```diff
@@ -0,0 +1,1713 @@
+
+pub struct AesCipher {
+    aes_object: AesObject,
+    key: Vec<u8>,
```
This is a plain `Vec`, which means that it won't be zeroed when dropped.
Consider either using `zeroize` or implement a `Drop` call to automatically free resources after usage.

helkoulak replied:
I have implemented a drop call for both ciphers to zero the key vectors.

gasbytes commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
```diff
@@ -0,0 +1,1713 @@
+        }
+    }
+
+    fn set_key(&self, key: &[u8]) -> Result<(), Error> {
```
This function takes `&self`, and sets the key on the FFI object, but` self.key` remains `None`, contrary to the `set_key` function of `AesCipher::set_key`.
Can you please also store raw bytes key here too for consistency?

helkoulak replied:
Done

gasbytes commented on .github/workflows/macos-build.yml:
```diff
@@ -56,6 +57,33 @@ jobs:
+          cd rustls
+          git fetch --tags
+          selected_tag=$(git tag -l "v/0\.23\.35")
+          git checkout "$selected_tag"
```
Maybe instead of using git tag -l followed by checkout, it might be easier to do a checkout directly like this:
`git checkout v/0.23.35`

julek-wolfssl replied:
No. Using the checkout action is the correct way.

helkoulak replied:
Done

gasbytes commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
```diff
@@ -0,0 +1,1713 @@
+mod tests {
+    use hex_literal::hex;
+    use rustls::crypto::tls13::HkdfExpander;
+    use std::prelude::rust_2015::ToString;
```
Is this necessary? ToString is included in teh standard prelude:
`https://doc.rust-lang.org/std/string/trait.ToString.html`

helkoulak replied:
It is unnecessary actually, you are right. I have removed it.

gasbytes commented on rustls-wolfcrypt-provider/Cargo.toml:
```diff
@@ -33,6 +33,7 @@ lazy_static = "1.5.0"
 hex-literal = "0.4.1"
 
 
+
```
Extra blank space.

helkoulak replied:
removed

gasbytes commented on rustls-wolfcrypt-provider/src/aead/quic.rs:
```diff
@@ -0,0 +1,1713 @@
+//!
+//! See draft-ietf-quic-tls.
+
+#![allow(clippy::type_complexity)]
```
Have you tried removing this to see if `clippy` comes back with any warnings/reports that might be worth fixing?

helkoulak replied:
By removing this lint, clippy will complain about a readability issue. It was fixed as seen in commit c0d2c73748fbd526f8e06fab4a4caac2f972bb73

gasbytes reviewed (changes requested):
Really nice work Hosam, I left a couple of comments.

gasbytes reviewed (dismissed):
Great work Hosam, LGTM.

anhu commented:
Contributor agreement signed, approved and filed. 
wolfSSL members, please proceed in the pull request process.

gasbytes commented:
@helkoulak please rebase your work against the latest main.
Thanks.

helkoulak commented:
> @helkoulak please rebase your work against the latest main. Thanks.

Will do ASAP. Thank you

helkoulak commented:
Branch has been rebased against the latest main

gasbytes committed:
- rustls test-suite modifications as local patch + fixtures under
.github/test-workspace
- run the rustls testsuite job against the pr's own provider replacing
external fork checkouts
